ui/pages/defi/components/defi-page.tsx

```diff
@@ -0,0 +1,231 @@
+import React from 'react';
+import { useHistory, useParams, Redirect } from 'react-router-dom';
+import { Hex } from '@metamask/utils';
+import {
+  Display,
+  FlexDirection,
+  IconColor,
+  TextColor,
+  TextVariant,
+} from '../../../helpers/constants/design-system';
+import {
+  Box,
+  ButtonIcon,
+  ButtonIconSize,
+  IconName,
+  SensitiveText,
+  Text,
+} from '../../../components/component-library';
+import { useI18nContext } from '../../../hooks/useI18nContext';
+import TokenCell from '../../../components/app/assets/token-cell';
+import { DEFAULT_ROUTE } from '../../../helpers/constants/routes';
+import { defiData } from './data';
+
+import { TokenWithFiatAmount } from '../../../components/app/assets/types';
+
+type Token = TokenWithFiatAmount & {
+  type: 'underlying' | 'underlying-claimable';
+};
+
+function extractUnderlyingTokens(positions: any, chainId: Hex): Token[][] {
```

Comment:
I know this is likely temporary but we should create a selector here and use `createDeepEqualSelector` to memoize the result.